### PR TITLE
Update: CSV File Processing and Uploading Done

### PR DESCRIPTION
**Pull request by mdhishaamakhtar**
The login routes are the same but the parsing and uploading of the csv files are done.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -39,7 +39,7 @@
 
 ## Contributors
 
-* [ < INSERT NAME HERE > ](INSERT_PROFILE_URL_HERE)
+* [ < Md Hishaam Akhtar > ](INSERT_PROFILE_URL_HERE)
 * [ < INSERT NAME HERE > ](INSERT_PROFILE_URL_HERE)
```
Comment (L04DB4L4NC3R) [inline README.md]: Remove all un-necessary placeholders

File: model/User.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
+const mongoose=require('mongoose');
+
+const userSchema=new mongoose.Schema({
+  name:{
+    type:String,
+    required: true,
+    min: 6,
```
Comment (L04DB4L4NC3R) [inline model/User.js]: What if my name is **Yash**. Remove the minimum size requirement

File: model/User.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  },
+  gender:{
+    type:String,
+    required:true
+  },
+  password:{
+    type:String,
+    required:true,
+    min:8,
+    max:1024
```
Comment (L04DB4L4NC3R) [inline model/User.js]: Why such a huge number for password?

File: package.json [hunk 4]
```diff
@@ -0,0 +1,24 @@
+{
+  "name": "auth",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "start": "nodemon index.js"
```
Comment (L04DB4L4NC3R) [inline package.json:7]: `npm start` has **nodemon** but dependencies do not

File: routes/auth.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  //VALIDATE
+  const {
+    error
+  } = registerValidation(req.body);
+  if (error) {
+    return res.status(400).send(error.details[0].message);
+  }
+
+  //CHECK FOR EXISTING email
+  const emailExists = await User.findOne({
```
Comment (L04DB4L4NC3R) [inline routes/auth.js:17]: Why not validate this in the `registerValidation` function?

File: routes/auth.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    email: req.body.email
+  });
+  if (emailExists) {
+    return res.status(400).send('Email already exists');
+  }
+
+  //HASH PASSWORDS
+  const salt = await bcrypt.genSalt(10);
```
Comment (L04DB4L4NC3R) [inline routes/auth.js]: Number of salt rounds should be hidden

File: routes/auth.js [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  const hashPassword = await bcrypt.hash(req.body.password, salt);
+
+  const user = new User({
+    name: req.body.name,
+    email: req.body.email,
+    password: hashPassword
+  });
+  try {
+    const savedUser = await user.save();
```
Comment (L04DB4L4NC3R) [inline routes/auth.js]: Use promises instead of try and catch here

File: routes/auth.js [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+  if (error) {
+    return res.status(400).send(error.details[0].message);
+  }
+
+  //CHECK FOR EXISTING email
+  const user = await User.findOne({
+    email: req.body.email
+  });
+  if (!user) {
+    return res.status(400).send('Email doesnt exist');
```
Comment (L04DB4L4NC3R) [inline routes/auth.js]: This should be `404`, not `400`

File: routes/auth.js [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+  }
+  //PASSWORD IS CORRECT
+  const validPass = await bcrypt.compare(req.body.password, user.password);
+  if (!validPass) {
+    return res.status(400).send('Password is wrong');
```
Comment (L04DB4L4NC3R) [inline routes/auth.js]: This should be `401` not `400`

File: routes/dashboard.js [hunk 10]
```diff
@@ -0,0 +1,13 @@
+const router = require('express').Router();
+const verify = require('./verifyToken');
+const User = require('../model/User');
+
+router.get('/', verify, async (req, res) => {
+  const use = await User.findOne({
+    _id: req.user
+  });
+  res.send(use);
```
Comment (L04DB4L4NC3R) [inline routes/dashboard.js]: Don't return the password of the user, even in hashed form

File: standalone-script/passwords.js [hunk 11]
```diff
@@ -0,0 +1,8 @@
+var uuid4 = require('uuid4');
+
+// Generate a new UUID
+const generate = () => {
```
Comment (L04DB4L4NC3R) [inline standalone-script/passwords.js]: This function is not needed

**Review by L04DB4L4NC3R — CHANGES_REQUESTED** [review]

File: routes/auth.js (hunk 5 above)
Comment (mdhishaamakhtar) [reply routes/auth.js:17]: Because registerValidation just returns if the credentials entered are valid or not using the @hapi/joi package whereaes this email exists finds in the database and returns.

Comment (L04DB4L4NC3R) [reply routes/auth.js:17]: > Because registerValidation just returns if the credentials entered are valid or not using the @hapi/joi package whereaes this email exists finds in the database and returns.

Okay, let it be

File: standalone-script/passwords.js [hunk 12]
```diff
@@ -2,8 +2,6 @@
 var uuid4 = require("uuid4");
 
 // Generate a new UUID
-const generate = () => {
+module.exports.generate = () => {
```
Comment (L04DB4L4NC3R) [inline standalone-script/passwords.js]: Why is this required? Just call the `uuid4()` function directly wherever needed

**Review by L04DB4L4NC3R — CHANGES_REQUESTED** [review]

File: README.md [hunk 13]
```diff
@@ -10,36 +10,42 @@
 
 
 ## Functionalities
-- [ ]  < Standalone Script for parsing csv file, generating passwords and uploading it to database >
-- [ ]  < insert functionality >
-- [ ]  < insert functionality >
-- [ ]  < insert functionality >
+- Standalone Script for parsing csv file, generating passwords and uploading it to database
```
Comment (L04DB4L4NC3R) [inline README.md]: Add the unticked check boxes here

File: README.md [hunk 14]
```diff
@@ -10,36 +10,42 @@
...
+	-  Node.js installed with npm enabled
+	-  An active internet connection to connect to the cloud database
+	-  All the npm packages mentioned in the package.json
 
-* < directions to install >
+* To setup the environment and install packages
 ```bash
-< insert code >
+npm init
+npm install @hapi/joi bcryptjs body-parser csv-parser csv-writer dotenv express mongoose nodemon fs uuid4 jsonwebtoken
```
Comment (L04DB4L4NC3R) [inline README.md]: No need. `package.json` has a list of packages to install. Just write `npm install`

File: README.md [hunk 15]
```diff
@@ -10,36 +10,42 @@
...
+cd standalone-script/
+node parse.js
 ```
 
 <br>
 
 ## Contributors
 
-* [ < Md Hishaam Akhtar > ](INSERT_PROFILE_URL_HERE)
+* [  Md Hishaam Akhtar  ]((https://github.com/mdhishaamakhtar))
```
Comment (L04DB4L4NC3R) [inline README.md]: Remove the extra pair of parenthesis

**Review by L04DB4L4NC3R — CHANGES_REQUESTED** [review]

**Review by L04DB4L4NC3R — APPROVED** [review]
